### PR TITLE
Start the RFC.md

### PR DESCRIPTION
**Pull request by waterlink**
This PR adds a simple example, trying to make a proposal for current open questions.

/cc @herrtreas

### CODE REVIEW

File: RFC.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+Then the service finds information about `docker-compose` in registry, such as:
+
+- Where and how to download the code (`tar`, `zip` and `git` url templates with
+  free `version` variable);
+- How to build code into a fully self-contained `docker` container (Github link
+  of form: `{https://github.com/heimathafen/recipes; /docker-compose}`);
+- How to run the final docker image (basically a script, that would imitate an
+  actual tool and abstract the fact that it is run inside of docker container,
+  i.e.: Github link of form `{https://github.com/heimathafen/recipes;
+  /docker-compose/run.sh}`);
```
Comment (waterlink) [inline RFC.md]: This bullets is an open question.

File: RFC.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+And the service builds the docker image for `docker-compose` of last released
+version and pushes it to `hub.docker.com` as
+`heimathafen/docker-compose:{{version}}` unless it was already built for this
```
Comment (waterlink) [inline RFC.md]: This is kind of an open question, if we should use something else?

File: RFC.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+version (or commit SHA);
+
+And the image gets downloaded to my machine;
+
+And the script to run the final docker image gets installed at
+`/opt/hafen/docker-compose/bin/docker-compose` and linked to
+`/usr/local/bin/docker-compose`,
```
Comment (waterlink) [inline RFC.md]: This is an open question, maybe one would want to have everything installed in `$HOME/.hafen` ? And so `hafen` installer will have to update `.bashrc` and `.zshrc` (and other popular shell `rc`s) to include `$HOME/.hafen/bin` to `PATH`.

File: RFC.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+And the service builds the docker image for `docker-compose` of last released
+version and pushes it to `hub.docker.com` as
+`heimathafen/docker-compose:{{version}}` unless it was already built for this
+version (or commit SHA);
+
+And the image gets downloaded to my machine;
+
+And the script to run the final docker image gets installed at
+`/opt/hafen/docker-compose/bin/docker-compose` and linked to
```
Comment (waterlink) [inline RFC.md]: This should probably be `/opt/hafen/docker-compose/{{version}}/bin/docker-compose`, i.e.: should include version to allow multiple versions being installed.

File: RFC.md (hunk 3 above)
Comment (waterlink) [inline RFC.md]: Additionally, about making a link to some folder on `PATH`, probably this should actually be a wrapper script, that depending on current situation can choose different version to run.

File: RFC.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+When I run this:
+
+```bash
+hafen install docker-compose
+```
+
+Then the service finds information about `docker-compose` in registry, such as:
+
+- Where and how to download the code (`tar`, `zip` and `git` url templates with
+  free `version` variable);
```
Comment (herrtreas) [inline RFC.md]: Free?

File: RFC.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+When I run this:
+
+```bash
+hafen install docker-compose
+```
+
+Then the service finds information about `docker-compose` in registry, such as:
+
+- Where and how to download the code (`tar`, `zip` and `git` url templates with
```
Comment (herrtreas) [inline RFC.md]: Shall we focus on purely cloning from Git at the beginning? Doing this we only need to handle `clone` and `checkout`

File: RFC.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  free `version` variable);
+- How to build code into a fully self-contained `docker` container (Github link
```
Comment (herrtreas) [inline RFC.md]: I was thinking of having an extra repo that keeps a list of receipts [just like home-brew does](https://github.com/Homebrew/homebrew/tree/master/Library/Formula).

A receipt could have three parts: build, test, run. I'd imagine this is contained by a Dockerfile, but right now this is highly spectacular and needs some prototyping. But at this point, we should agree on where the registry is getting the stuff from and `build,test,run` is to be decided later and probably deserves its own PR

wdyt?

File: RFC.md (hunk 2 above)
Comment (herrtreas) [inline RFC.md]: Yeah we could, there is plenty of options as long as we base this on Docker Registry (which we probably should). Amazon is going to offer a Registry service as well soon or we could run our own (which is my least preferred option tbh)

File: RFC.md (hunk 3 above)
Comment (herrtreas) [inline RFC.md]: > This is an open question, maybe one would want to have everything installed in $HOME/.hafen ? And so hafen installer will have to update .bashrc and .zshrc (and other popular shell rcs) to include $HOME/.hafen/bin to PATH.

Yeah. I'd love to keep this all in a know intrusive folder like `~/.hafen`

> Additionally, about making a link to some folder on PATH, probably this should actually be a wrapper script, that depending on current situation can choose different version to run.

This is about versioning right? Shall we have a dedicated `rfc` labeled issue based discussion about it?

File: RFC.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+`/opt/hafen/docker-compose/bin/docker-compose` and linked to
+`/usr/local/bin/docker-compose`,
+
+- Given I already have something in `/usr/local/bin/docker-compose`;
+- And it is not a link to `/opt/hafen/*/bin/*`;
+- Then `hafen` asks me a (yes/no) question if I want to override it;
+
+When I run `docker-compose --version`;
+
+Then I see a latest released version of `docker-compose`.
```
Comment (herrtreas) [inline RFC.md]: .. and some info that it is wrapped by Heimathafen, containing links to Issues and stuff

File: RFC.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+And the image gets downloaded to my machine;
+
+And the script to run the final docker image gets installed at
+`/opt/hafen/docker-compose/bin/docker-compose` and linked to
+`/usr/local/bin/docker-compose`,
+
+- Given I already have something in `/usr/local/bin/docker-compose`;
+- And it is not a link to `/opt/hafen/*/bin/*`;
+- Then `hafen` asks me a (yes/no) question if I want to override it;
```
Comment (herrtreas) [inline RFC.md]: Yes. I think the corresponding concept in Homebrew is `brew link`

File: RFC.md (hunk 5 above)
Comment (waterlink) [inline RFC.md]: `"https://github.com/user/repo/archives/{{version}}.zip"`

File: RFC.md (hunk 8 above)
Comment (waterlink) [inline RFC.md]: Agree

File: RFC.md (hunk 9 above)
Comment (waterlink) [inline RFC.md]: Well, this is pretty much obsolete, if we are going with `~/.hafen` approach?

File: RFC.md (hunk 2 above)
Comment (waterlink) [inline RFC.md]: I think, that using `hub.docker.com` can be the easiest solution for now.

Comment (waterlink) [inline RFC.md]: We just need to remember, that at any point of time, we should be able to substitute it with private/aws registry.

File: RFC.md (hunk 3 above)
Comment (waterlink) [inline RFC.md]: > Yeah. I'd love to keep this all in a know intrusive folder like ~/.hafen

Ok. In that case would you hijack `.bashrc`, `.zshrc` (and friends) automatically, or would you ask user to add certain commands themselves?

> This is about versioning right? Shall we have a dedicated rfc labeled issue based discussion about it?

Sounds good.

File: RFC.md (hunk 6 above)
Comment (waterlink) [inline RFC.md]: Yeah, sounds good to me.

File: RFC.md (hunk 7 above)
Comment (waterlink) [inline RFC.md]: It actually can be as simple as just getting all the stuff from Github, something like: `https://github.com/heimathafen/recipes` or `https://github.com/heimathafen/library` ?

File: RFC.md (hunk 5 above)
Comment (waterlink) [inline RFC.md]: Might be not the best word, but that is how you would say in Math I think..

Comment (waterlink) [inline RFC.md]: Do you have a suggestion for a better name for a "free variable `version`" in that meaning and context ?

Comment (waterlink) [inline RFC.md]: Oh, actually, in context of doing only `git`, we need no templates, no free variables there.

**Comment by waterlink** [comment]
@herrtreas I have incorporated your feedback, can you take another look/read ?

**Comment by herrtreas** [comment]
@waterlink Yep, I guess thats it! :tada:

**Pull request merged**